File: src/main/java/com/yupi/template/guardrail/PromptGuardrail.java
package com.yupi.template.guardrail;

import com.yupi.template.exception.BusinessException;
import com.yupi.template.exception.ErrorCode;

import java.util.List;
import java.util.regex.Pattern;

/**
 * Prompt 安全审查护轨：长度、敏感词、注入模式检测
 *
 * @author <a href="https://codefather.cn">编程导航学习圈</a>
 */
public final class PromptGuardrail {

    private static final int MAX_PROMPT_LENGTH = 4000;

    private static final List<String> SENSITIVE_WORDS = List.of(
            "忽略之前的指令", "ignore previous instructions", "ignore above", "ignore all",
            "破解", "hack", "绕过", "bypass", "越狱", "jailbreak",
            "无视", "disregard", "forget everything"
    );

    private static final List<Pattern> INJECTION_PATTERNS = List.of(
            Pattern.compile("(?i)ignore\\s+(?:previous|above|all)\\s+(?:instructions?|commands?|prompts?)"),
            Pattern.compile("(?i)(?:forget|disregard)\\s+(?:everything|all)\\s+(?:above|before)"),
            Pattern.compile("(?i)(?:pretend|act|behave)\\s+(?:as|like)\\s+(?:if|you\\s+are)"),
            Pattern.compile("(?i)system\\s*:\\s*you\\s+are"),
            Pattern.compile("(?i)new\\s+(?:instructions?|commands?|prompts?)\\s*:")
    );

    private PromptGuardrail() {
    }

    /**
     * 校验 prompt，不通过则抛出 BusinessException
     */
    public static void validate(String prompt) {
        if (prompt == null) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR, "输入内容不能为空");
        }
        String t = prompt.trim();
        if (t.isEmpty()) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR, "输入内容不能为空");
        }
        if (t.length() > MAX_PROMPT_LENGTH) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR,
                    "输入内容过长，请勿超过 " + MAX_PROMPT_LENGTH + " 字");
        }
        String lower = t.toLowerCase();
        for (String w : SENSITIVE_WORDS) {
            if (lower.contains(w.toLowerCase())) {
                throw new BusinessException(ErrorCode.PARAMS_ERROR, "输入包含不当内容，请修改后重试");
            }
        }
        for (Pattern p : INJECTION_PATTERNS) {
            if (p.matcher(t).find()) {
                throw new BusinessException(ErrorCode.PARAMS_ERROR, "检测到恶意输入，请求被拒绝");
            }
        }
    }
}

File: python-backend/app/utils/prompt_guardrail.py
"""
Prompt 安全审查护轨：长度、敏感词、注入模式检测
@author <a href="https://codefather.cn">编程导航学习圈</a>
"""
import re
from typing import List, Pattern

from app.core.errors import BusinessException, ErrorCode

MAX_PROMPT_LENGTH = 4000

SENSITIVE_WORDS: List[str] = [
    "忽略之前的指令", "ignore previous instructions", "ignore above", "ignore all",
    "破解", "hack", "绕过", "bypass", "越狱", "jailbreak",
    "无视", "disregard", "forget everything"
]

INJECTION_PATTERNS: List[Pattern] = [
    re.compile(r"(?i)ignore\s+(?:previous|above|all)\s+(?:instructions?|commands?|prompts?)"),
    re.compile(r"(?i)(?:forget|disregard)\s+(?:everything|all)\s+(?:above|before)"),
    re.compile(r"(?i)(?:pretend|act|behave)\s+(?:as|like)\s+(?:if|you\s+are)"),
    re.compile(r"(?i)system\s*:\s*you\s+are"),
    re.compile(r"(?i)new\s+(?:instructions?|commands?|prompts?)\s*:")
]


def validate(prompt: str) -> None:
    """
    校验 prompt，不通过则抛出 BusinessException

    Args:
        prompt: 用户输入的提示词

    Raises:
        BusinessException: 校验不通过时
    """
    if prompt is None:
        raise BusinessException(ErrorCode.PARAMS_ERROR, "输入内容不能为空")
    text = prompt.strip()
    if not text:
        raise BusinessException(ErrorCode.PARAMS_ERROR, "输入内容不能为空")
    if len(text) > MAX_PROMPT_LENGTH:
        raise BusinessException(
            ErrorCode.PARAMS_ERROR,
            f"输入内容过长，请勿超过 {MAX_PROMPT_LENGTH} 字"
        )
    lower = text.lower()
    for word in SENSITIVE_WORDS:
        if word.lower() in lower:
            raise BusinessException(ErrorCode.PARAMS_ERROR, "输入包含不当内容，请修改后重试")
    for pattern in INJECTION_PATTERNS:
        if pattern.search(text):
            raise BusinessException(ErrorCode.PARAMS_ERROR, "检测到恶意输入，请求被拒绝")
